### ISSUE TITLE
"Body discarded during HAR generation: longer than limit of 500000 bytes"

### ISSUE THREAD

**Issue by lord-ne** [httptoolkit/httptoolkit#477]
### Repro steps

Responses containing things like images that are too large will not be saved to the HAR when saving all response in a session to a HAR.

* Intercept an exchange containing a large image
* Click "export these exchanges as a HAR file" button
* HAR does not contain image, has a comment indicating it was not saved

Saving an individual exchange to a HAR file does seem to include the image content. The GUI also shows the image.

### How often does this bug happen?

Every time

### The desktop OS you're using

Windows 11

### Details of other apps/devices

Example URL with a too-large image: https://upload.wikimedia.org/wikipedia/commons/8/85/%22_Shot_From_The_Sky%22_Army_Show_1945_Oak_Ridge_%2824971013612%29.jpg

### Error screenshot

Shows up correctly in GUI: 
![image](https://github.com/httptoolkit/httptoolkit/assets/34069872/1c373cbf-0b51-46f2-a683-ed57d7d4ab8c)

But is not in the saved HAR:

```
    "content": {
        "mimeType": "image/jpeg",
        "size": 6268505,
        "comment": "Body discarded during HAR generation: longer than limit of 500000 bytes"
    },
```

**Comment by pimterry** [comment]
Yes, this is by design. If you export a large volume of traffic, for most use cases it can be a problem if that JSON file becomes too large.

HAR files aren't a great format for larger binary data like this, so it's usually better to download the body directly instead. You can download bodies (request or response) directly from the download icon shown in the request/response body section, which will give you the content in its raw form, as a single file without the HAR wrapper.

Can you explain more about your context, and the use case where you do want to be able to batch export large data like this? I think HAR might not be the right format for you, but I'm open to alternatives (for example, maybe we should allow batch downloading request/response bodies as separate files into a folder?).

**Comment by lord-ne** [comment]
Use case is saving a set of images sent to an app

**Comment by lord-ne** [comment]
That would be enough for me to work with, although it's a bit inconvenient because I had been using a Python script to loop over the responses and do other things with the image URL (for example, I was downloading different qualities of the images by modifying the URLs). I'm able to work around this issue anyway by just having the Python script re-download the original image I guess.

Personally I would prefer just having a checkbox (off by default) in the settings to allow saving larger HAR files. Even if it can cause problems, people can evaluate their own use case to decide what they need. Alternatively, just limit the total size of the HAR file rather than of individual bodies; my sessions for example are very short (less than 100 requests), so the overall HAR file size is not crazy. Also, a warning popup when saving truncates something would also be nice.

**Comment by pimterry** [comment]
Thanks for the context, that makes sense. I think you might also find that #37 would be a useful solution to this - in that cases you could automatically process these requests individually with custom logic, as they're proxied. You can do this today by using a custom Mockttp setup as an upstream server (more info here: https://httptoolkit.com/blog/javascript-mitm-proxy-mockttp/).

I'll keep this issue open for now anyway as I look into it more. I do think being able to batch export files is likely to be the best approach here, as including more than a handful of large bodies like this in a HAR file will very quickly make it unusable with most HAR tools, and I really don't want to start exporting unusable content. And adding options for features that are generally a bad idea is usually a quick way to end up with a very unusable product.

With the proposed multi-select, it could be possible to export both the HAR and the set of response bodies at the same time though (exporting HAR is definitely something that will be supported by that multi-select setup) so you would still be able to do what you're suggesting - export all bodies as files in a folder, export all traffic metadata to a single HAR, and then just go through the HAR and read the corresponding saved body files from the folder if it's not in the HAR already.
